bindings/rust/wasmedge-sys/src/instance/function.rs

```diff
@@ -0,0 +1,162 @@
+use crate::{wasmedge, Value, WasmFnIO, HOST_FUNCS};
+
+use core::ffi::c_void;
+use std::convert::TryInto;
+
+use rand::Rng;
+
+extern "C" fn wraper_fn(
+    key_ptr: *mut c_void,
+    _data: *mut c_void,
+    _mem_cxt: *mut wasmedge::WasmEdge_MemoryInstanceContext,
+    params: *const wasmedge::WasmEdge_Value,
+    param_len: u32,
+    returns: *mut wasmedge::WasmEdge_Value,
+    return_len: u32,
+) -> wasmedge::WasmEdge_Result {
+    let key = key_ptr as *const usize as usize;
+    let mut result: Result<Vec<Value>, u8> = Err(0);
+
+    let mut input: Vec<Value> = {
+        let raw_input = unsafe {
+            std::slice::from_raw_parts(
+                params,
+                param_len
+                    .try_into()
+                    .expect("len of params should not greater than usize"),
+            )
+        };
+        raw_input.iter().map(|r| (*r).into()).collect()
+    };
+    input.remove(0);
+
+    let return_len = return_len
+        .try_into()
+        .expect("len of returns should not greater than usize");
+    let raw_returns = unsafe { std::slice::from_raw_parts_mut(returns, return_len) };
+
+    HOST_FUNCS.with(|f| {
+        let mut host_functions = f.borrow_mut();
+        let real_fn = host_functions
+            .remove(&key)
+            .expect("host function should there");
+        result = real_fn(input);
+    });
+
+    match result {
+        Ok(v) => {
+            assert!(v.len() == return_len);
+            for (idx, item) in v.into_iter().enumerate() {
+                raw_returns[idx] = item.into();
+            }
+            wasmedge::WasmEdge_Result { Code: 0 }
+        }
+        Err(c) => wasmedge::WasmEdge_Result { Code: c as u8 },
+    }
+}
+
+#[derive(Debug)]
+pub struct Function {
+    pub(crate) ctx: *mut wasmedge::WasmEdge_FunctionInstanceContext,
+    pub(crate) registed: bool,
+    ty: Option<Type>,
+}
+
+impl Function {
+    /// wasmedge::WasmEdge_FunctionInstanceCreate take C functions
+    /// This may not be implement
+    pub fn create<I: WasmFnIO, O: WasmFnIO>(
+        _f: Box<dyn std::ops::Fn(Vec<Value>) -> Vec<Value>>,
+    ) -> Self {
+        unimplemented!()
+    }
+
+    // TODO:
+    // binding errors and restict the error types
+    #[allow(clippy::type_complexity)]
+    /// Binding Rust function (HostFunction) to WasmEdgeFunction
+    pub fn create_bindings<I: WasmFnIO, O: WasmFnIO>(
+        real_fn: Box<dyn Fn(Vec<Value>) -> Result<Vec<Value>, u8>>,
+    ) -> Self {
+        let mut key = 0usize;
+        HOST_FUNCS.with(|f| {
+            let mut rng = rand::thread_rng();
```

Comment:
Do we need a rng number here? Or we can have a local counter?

Reply:
The counter will become another global variable, once we have more threads or processes, there will be a lock on it.  This will be nice and lock-free in the far future without changing code, so the `rng` is decided to use here.

Reply:
Sure. Using `rng` + hashmap is fine.